types/proposal.go

```diff
@@ -84,15 +84,20 @@ func (p *Proposal) ValidateBasic() error {
 // configured Precision and MsgDelay parameters.
 // Specifically, a proposed block timestamp is considered timely if it is satisfies the following inequalities:
 //
-// proposedBlockTime > validatorLocaltime - Precision && proposedBlockTime < validatorLocalTime + Precision + MsgDelay.
+// localtime >= proposedBlockTime - Precision
+// localtime <= proposedBlockTime + MsgDelay + Precision
 //
 // For more information on the meaning of 'timely', see the proposer-based timestamp specification:
 // https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
-func (p *Proposal) IsTimely(clock tmtime.Source, tp TimingParams) bool {
-	lt := clock.Now()
-	lhs := lt.Add(-tp.Precision)
-	rhs := lt.Add(tp.Precision).Add(tp.MessageDelay)
-	if lhs.Before(p.Timestamp) && rhs.After(p.Timestamp) {
+func (p *Proposal) IsTimely(recvTime time.Time, tp TimingParams, initialHeight int64) bool {
+	// lhs is `proposedBlockTime - Precision` in the first inequality
+	lhs := p.Timestamp.Add(-tp.Precision)
+	// rhs is `proposedBlockTime + MsgDelay + Precision` in the second inequality
+	rhs := p.Timestamp.Add(tp.MessageDelay).Add(tp.Precision)
+
+	recvTimeAfterOrEqLHS := recvTime.After(lhs) || recvTime.Equal(lhs)
```

Comment:
It seems the code would be easier to read if we had `<` instead of `<=` inequalities. In fact, the specification is kind of imprecise regarding this point. I would prefer open intervals, but I'm going to discuss that with Josef.

Reply:
Agreed. I could also flip the logic to:
```
	if recvTime.Before(lhs) || (p.Height != initialHeight) && recvTime.After(rhs) {
		return false
	}
        return true
```